### PR TITLE
database-setup-bug-fixes 

### PR DESCRIPTION
**Pull request by Grant-Postma**
Touch Listener hitbox correction.
Change back to MainMenu if game_uuid doesn't exist.
Added game over firebase field in game_uuid.
Removeing Game_Uuid on game over.
Added game_outcome tracker.
Leave button functionality.

### CODE REVIEW

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp [hunk 1]
```diff
@@ -326,32 +331,44 @@ TicTacToeLayer::TicTacToeLayer(string game_uuid) {
...
+    // initialization to failed.
+    auto future_game_uuid =
+        database->GetReference("game_data").Child(join_game_uuid).GetValue();
+    WaitForCompletion(future_game_uuid, "GetGameDataMap");
+    auto game_uuid_snapshot = future_game_uuid.result();
+    if (!game_uuid_snapshot->value().is_map()) {
+      initialization_failed = true;
+    } else {
+      ref = database->GetReference("game_data").Child(join_game_uuid);
+      auto fIncrementTotalUsers = ref.RunTransaction([](MutableData* data) {
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeLayer.cpp:351]: nit: snake case

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp [hunk 2]
```diff
@@ -365,6 +382,47 @@ TicTacToeLayer::TicTacToeLayer(string game_uuid) {
...
+  leave_button_sprite_touch_listener->onTouchBegan =
+      [this](Touch* touch, Event* event) -> bool {
+    auto bounds = event->getCurrentTarget()->getBoundingBox();
+    auto point = touch->getLocation();
+    // Replaces the scene with a new TicTacToe scene if the touched point is
+    // within the bounds of the button.
+    if (bounds.containsPoint(point)) {
+      // Update the game_outcome to reflect if the you rage quit or left
+      // pre-match.
+      if (remaining_tiles.size() == kNumberOfTiles)
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeLayer.cpp:408]: I don't think the style guide explicitly requires it here, but I prefer to have braces around all my if/else statements.  Unbraced if bodies make me nervous.

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp [hunk 3]
```diff
@@ -501,11 +558,28 @@ void TicTacToeLayer::update(float /*delta*/) {
...
+      // Set game_outcome to reflect the game ended in a tie.
+      game_outcome = kGameTied;
+    }
+  }
+  // Shows the end game label for kEndGameFramesMax to show the result of the
+  // game.
+  else if (game_over_listener->got_value()) {
+    if (game_outcome == kGameDisbanded &&
+        remaining_tiles.size() != kNumberOfTiles)
+      game_outcome = kGameWon;
```
Comment (alexames) [inline demos/TicTacToe/Classes/TicTacToeLayer.cpp:573]: Braces here too plz

File: demos/TicTacToe/Classes/MainMenuScene.cpp [hunk 4]
```diff
@@ -22,7 +21,8 @@ bool MainMenuScene::init() {
   // Creates a sprite for the create button and sets its position to the center
   // of the screen. TODO(grantpostma): Dynamically choose the location.
   auto create_button = Sprite::create("create_game.png");
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/MainMenuScene.cpp:23]: Aren't the file names in other files defined as a constant?

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp [hunk 5]
```diff
@@ -33,6 +33,15 @@ static const int kEmptyTile = -1;
 static const int kPlayerOne = 0;
 static const int kPlayerTwo = 1;
 static const int kNumberOfPlayers = 2;
+
+// End game outcomes.
+static const int kGameWon = 0;
+static const int kGameLost = 1;
+static const int kGameTied = 2;
+static const int kGameDisbanded = 3;
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeLayer.cpp:41]: Should these be an enumerated type?

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp [hunk 6]
```diff
@@ -436,20 +500,20 @@ TicTacToeLayer::TicTacToeLayer(string game_uuid) {
...
       awaiting_opponenet_move = true;
       waiting_label->setString("waiting");
       if (GameOver(board)) {
         // Set game_over_label to reflect the use won.
         game_over_label->setString("you won!");
       } else if (remaining_tiles.size() == 0) {
-        // Set game_over_label to reflect the game ended in a tie.
-        game_over_label->setString("you tied.");
-        // Changing back to the main menu scene.
+        // Update game_outcome to reflect the use tied.
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeLayer.cpp:514]: "user" instead of "use" ?

File: demos/TicTacToe/Classes/TicTacToeLayer.h [hunk 7]
```diff
@@ -35,23 +36,36 @@ class TicTacToeLayer : public Layer {
   TicTacToeLayer(std::string);
   ~TicTacToeLayer();
   virtual void TicTacToeLayer::update(float);
+  // Tracks whether the board was unable to build
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeLayer.h:39]: nit: add a period

File: demos/TicTacToe/Classes/TicTacToeLayer.h [hunk 8]
```diff
@@ -35,23 +36,36 @@ class TicTacToeLayer : public Layer {
...
+  bool initialization_failed = false;
+  // Tracks the game outcome
```
Comment (DellaBitta) [inline demos/TicTacToe/Classes/TicTacToeLayer.h:41]: nit: add a period

**Comment by DellaBitta** [comment]
Thanks for the changes!

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp (hunk 3 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeLayer.cpp:578]: 789ee14 should address this.

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp (hunk 2 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeLayer.cpp]: 789ee14 should address this.

File: demos/TicTacToe/Classes/TicTacToeLayer.cpp (hunk 1 above)
Comment (Grant-Postma) [reply demos/TicTacToe/Classes/TicTacToeLayer.cpp]: 789ee14 should address this.

**Pull request merged**